streams-contrib/streams-provider-moreover/src/main/java/org/apache/streams/data/moreover/MoreoverProviderTask.java

```diff
@@ -18,77 +18,168 @@
+    }
+
+    /**
+     * Polls the moreover api till it returns zero results.  If it is in perpetual mode it will then continue to poll on a
+     * cadence.  If not in perpetual mode it will exit the method.
+     */
     @Override
     public void run() {
-        while(true) {
+        this.running.set(true);
+        try {
+            // have to poll once to ensure initial sequence id
+            MoreoverClient moreoverClient = getMoreoverClient(this.apiId, this.apiKey, this.lastSequence);
+            MoreoverResult initialResult = moreoverClient.getNextBatch();
+            LOGGER.debug("Adding {} articles to the output queue.", initialResult.numArticles());
+            this.lastSequence = initialResult.getMaxSequencedId().toString();
+            addMoreoverArticlesToOutputQueue(initialResult);
+            pollTillClientReturnsLessThanMaxResults(moreoverClient);
+            if (this.perpetualMode) {
+                perpetualRun(moreoverClient);
+            }
+        } catch (IOException ioe) {
+            LOGGER.error("Unable to initialize moreover data with the api. apiId={} apiKey={} lastSequence={}", this.apiId, this.apiKey, this.lastSequence);
+            LOGGER.error("Unable to initialize moreover data : {}", ioe);
+        } catch (InterruptedException ie) {
+            LOGGER.warn("Caught InteruptedException. Shutting down.", ie);
+            Thread.currentThread().interrupt();
+        } finally {
+            this.running.set(false);
+        }
+    }
+
+    /**
+     * Polls till the moreover client returns less than max result
+     * @param client
+     * @throws InterruptedException
+     */
+    protected void pollTillClientReturnsLessThanMaxResults(MoreoverClient client) throws InterruptedException {
+        int resultSize = 0;
+        int attempt = 0;
+        do {
             try {
-                ensureTime(moClient);
-                MoreoverResult result = moClient.getArticlesAfter(lastSequence, 500);
-                started = true;
-                lastSequence = result.process().toString();
-                for(StreamsDatum entry : ImmutableSet.copyOf(result.iterator()))
-                    results.offer(entry);
-                logger.info("ApiKey={}\tlastSequenceid={}", this.apiKey, lastSequence);
-
-            } catch (Exception e) {
-                logger.error("Exception while polling moreover", e);
+                ensureTime(client);
+                MoreoverResult result = client.getNextBatch();
+                LOGGER.debug("Adding {} articles to the output queue.", result.numArticles());
+                addMoreoverArticlesToOutputQueue(result);
+                resultSize = result.numArticles();
+                attempt = 0; //reset attempt
+            } catch (IOException ioe) {
+                ++attempt;
+                if(attempt >= MAX_ATTEMPTS) {
+                    LOGGER.error("Unable to get api response after {} attempts. Shutting down", attempt);
```

Comment:
This log says that we're shutting down but i don't see anything that will explicitly do so.